### PR TITLE
ER-452 Clone Vacancy

### PR DESCRIPTION
**Pull request by leewadhams**
- Clones an existing vacancy.
- Training details are now part of the dashboard projection.
- A new job to re-create all dashboards has been added to handle the change to the dashboard projection (queue: **generate-all-employer-dashboard-queue**).
- ~~Question: Should we have the release process add the message onto the queue when needed or leave to be manual initially?~~ Adding a release step to regenerate on release

### CODE REVIEW

File: src/Employer/Employer.Web/ViewModels/CreateVacancyOptions/CreateVacancyOptionsEditModel.cs [hunk 1]
```diff
@@ -0,0 +1,6 @@
+﻿namespace Esfa.Recruit.Employer.Web.ViewModels.CreateVacancyOptions
+{
+    public class CreateVacancyOptionsEditModel : CreateVacancyOptionsViewModel
```
Comment (chrisjensenuk) [inline src/Employer/Employer.Web/ViewModels/CreateVacancyOptions/CreateVacancyOptionsEditModel.cs]: Is this inheritance the right way round?

File: src/Shared/Recruit.Vacancies.Client/Application/CommandHandlers/CloneVacancyCommandHandler.cs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            clone.CreatedDate = now;
+            clone.LastUpdatedByUser = message.User;
+            clone.LastUpdatedDate = now;
+            clone.SourceOrigin = SourceOrigin.EmployerWeb;
+            clone.SourceType = SourceType.Clone;
+            clone.SourceVacancyReference = vacancy.VacancyReference;
+            clone.Status = VacancyStatus.Draft;
+            clone.IsDeleted = false;
+
+            // Propertis to remove
```
Comment (chrisjensenuk) [inline src/Shared/Recruit.Vacancies.Client/Application/CommandHandlers/CloneVacancyCommandHandler.cs]: typo

File: src/Shared/Recruit.Vacancies.Client/Application/CommandHandlers/CloneVacancyCommandHandler.cs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+            // Properties to replace
+            clone.Id = message.NewVacancyId;
+            clone.CreatedByUser = message.User;
+            clone.CreatedDate = now;
+            clone.LastUpdatedByUser = message.User;
+            clone.LastUpdatedDate = now;
+            clone.SourceOrigin = SourceOrigin.EmployerWeb;
+            clone.SourceType = SourceType.Clone;
+            clone.SourceVacancyReference = vacancy.VacancyReference;
```
Comment (chrisjensenuk) [inline src/Shared/Recruit.Vacancies.Client/Application/CommandHandlers/CloneVacancyCommandHandler.cs]: Should we trap the source Vacancy Id guid?

File: src/Employer/Employer.Web/ViewModels/CreateVacancyOptions/CreateVacancyOptionsEditModel.cs (hunk 1 above)
Comment (leewadhams) [reply src/Employer/Employer.Web/ViewModels/CreateVacancyOptions/CreateVacancyOptionsEditModel.cs]: Nope!!

File: src/Shared/Recruit.Vacancies.Client/Application/CommandHandlers/CloneVacancyCommandHandler.cs (hunk 2 above)
Comment (leewadhams) [reply src/Shared/Recruit.Vacancies.Client/Application/CommandHandlers/CloneVacancyCommandHandler.cs]: 👍

File: src/Employer/Employer.Web/ViewModels/CreateVacancyOptions/CreateVacancyOptionsViewModel.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+using System.ComponentModel.DataAnnotations;
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+using Microsoft.AspNetCore.Mvc.Rendering;
+
+namespace Esfa.Recruit.Employer.Web.ViewModels.CreateVacancyOptions
+{
+    public class CreateVacancyOptionsViewModel 
+    {
+        [Required(ErrorMessage = ValidationMessages.CreateVacancyOptionsConfirmationMessages.SelectionRequired)]
+        public Guid? VacancyId { get; set; }
```
Comment (sachpatel) [inline src/Employer/Employer.Web/ViewModels/CreateVacancyOptions/CreateVacancyOptionsViewModel.cs]: We shouldn't be posting a a view model should we?

File: src/Employer/Employer.Web/ViewModels/CreateVacancyOptions/CreateVacancyOptionsEditModel.cs (hunk 1 above)
Comment (sachpatel) [reply src/Employer/Employer.Web/ViewModels/CreateVacancyOptions/CreateVacancyOptionsEditModel.cs]: there shouldn't be any inheritance and are we not posting all input values as strings and then using type coersion??

File: src/Shared/Recruit.Vacancies.Client/Application/CommandHandlers/CloneVacancyCommandHandler.cs (hunk 3 above)
Comment (leewadhams) [reply src/Shared/Recruit.Vacancies.Client/Application/CommandHandlers/CloneVacancyCommandHandler.cs]: Do we need to if we've got the VacancyReference? The conversation with @mgwilliam around regeneration of documents relied on the fact that we don't have store internal identifiers.

File: src/Employer/Employer.Web/ViewModels/CreateVacancyOptions/CreateVacancyOptionsViewModel.cs (hunk 4 above)
Comment (leewadhams) [reply src/Employer/Employer.Web/ViewModels/CreateVacancyOptions/CreateVacancyOptionsViewModel.cs]: ??

File: src/Shared/Recruit.Vacancies.Client/Application/CommandHandlers/CloneVacancyCommandHandler.cs (hunk 3 above)
Comment (chrisjensenuk) [reply src/Shared/Recruit.Vacancies.Client/Application/CommandHandlers/CloneVacancyCommandHandler.cs]: Ignore me. For some reason I thought we would lose track of which 'version' of a vacancy was used for the clone but of cause the clone gets a unique VacancyReference.

**Pull request merged**